courseScheduleII.js

```diff
@@ -0,0 +1,110 @@
+/*
+For this approach, we can represent the input as a graph where in order to take a course, 
+you have to complete a prereq, which is the same as two nodes conntected by an edge.
+ 
+In order to traverse to the next node, you have to start with the one connected to it prior. 
+As you can see, it's going to be a directed graph, since we can also travel in one direction.
+
+Since we are trying to find a valid ordering for all these courses that respects our prereqs, 
+we can approach it the same way that processes all of ours nodes that respects the ordering of ours edges.
+Thus we can use a BFS that allows use a top sort algorithm.
+*/
+
+/*
+Question:
+There are a total of numCourses courses you have to take, labeled from 0 to numCourses - 1. 
+You are given an array prerequisites where prerequisites[i] = [ai, bi] indicates that you must 
+take course bi first if you want to take course ai.
+
+For example, the pair [0, 1], indicates that to take course 0 you have to first take course 1.
+Return the ordering of courses you should take to finish all courses. 
+If there are many valid answers, return any of them. 
+
+If it is impossible to finish all courses, return an empty array.
+*/
+
+const { Queue } = require('./utils/queue');
+
+const buildGraph = (numNodes, prereqNodes) => {
+    const graph = {};
+
+    for (let node = 0; node < numNodes; node++) {
+        graph[node] = [];
+    }
+
+    for (const prereqNode of prereqNodes) {
+        const [mainCource, prereq] = prereqNode;
+
+        graph[prereq].push(mainCource);
+    }
+
+    return graph;
+};
+
+const getIndegreeMap = (numNodes, prereqNodes) => {
+    const indegreeMap = {};
+
+    for (let node = 0; node < numNodes; node++) {
+        indegreeMap[node] = 0;
+    }
+
+    for (const prereqNode of prereqNodes) {
+        const [node, prereq] = prereqNode;
+        indegreeMap[node]++;
+    }
+
+    return indegreeMap;
+};
+
+const getIndgreeZeroNodesQueue = (indegreeMap, numNodes) => {
+    const indegreeZeroNodesQueue = new Queue();
+
+    for (let node = 0; node < numNodes; node++) {
+        if (indegreeMap[node] === 0) indegreeZeroNodesQueue.enqueue(node);
+    }
+
+    return indegreeZeroNodesQueue;
+};
+
+const getValidOrderNodes = (graph, indegreeMap, numNodes) => {
+    const indegreeZeroNodesQueue = getIndgreeZeroNodesQueue(indegreeMap, numNodes);
```

Comment:
P: Why do we care about a node's indegree? What does it represent in the context of this problem? Also, what does it mean (in the context of this problem) if a node has indegree 0?

Reply:
We care about the node's indegree because it gives us the number of inward directed graph edges associated with that node. For this problem, it represents the prereqs required to take a particular class. For example, class A requires the prereq class B, so class A has an indegree of 1. If a node has an indegree of 0, it means there are no inward edge for that node.

Reply:
You're correct about what indegree represents in the context of this problem. You did not answer what it means if a node has indegree 0 **in the context of this problem**.

Reply:
For the context of this problem, we care about the node being an indegree 0 (when there are no more prereqs to take) because that's when we can push it into our valid sorting array. Thus, gives us a valid topological sort and the ideal ordering of classes to take.

Reply:
Yes, a node with indegree 0 is a class that we are ready to take (we have satisfied the prereqs).